File: sdk/python/urlshortener/types/qr.py
"""QR code related type definitions."""

from typing import List, Optional, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field, HttpUrl, validator


class QRFormat(str, Enum):
    """Available QR code formats."""
    PNG = "png"
    SVG = "svg"
    JPEG = "jpeg"
    PDF = "pdf"


class QRErrorCorrectionLevel(str, Enum):
    """QR code error correction levels."""
    LOW = "L"      # ~7% correction
    MEDIUM = "M"   # ~15% correction
    QUARTILE = "Q" # ~25% correction
    HIGH = "H"     # ~30% correction


class QRCodeOptions(BaseModel):
    """QR code generation options."""
    size: Optional[int] = Field(None, ge=50, le=2000, description="QR code size in pixels")
    format: Optional[QRFormat] = Field(QRFormat.PNG, description="Output format")
    level: Optional[QRErrorCorrectionLevel] = Field(QRErrorCorrectionLevel.MEDIUM, description="Error correction level")
    margin: Optional[int] = Field(None, ge=0, le=50, description="Margin size")
    dark_color: Optional[str] = Field(None, description="Dark color (hex code)")
    light_color: Optional[str] = Field(None, description="Light color (hex code)")
    
    @validator('dark_color', 'light_color')
    def validate_color(cls, v):
        """Validate color hex codes."""
        if v is not None:
            import re
            if not re.match(r'^#[0-9A-Fa-f]{6}$', v):
                raise ValueError('Color must be a valid hex code (e.g., #000000)')
        return v


class QRCodeResponse(BaseModel):
    """QR code generation response."""
    data: str = Field(description="Base64 encoded image or SVG string")
    format: QRFormat = Field(description="Output format")
    size: int = Field(description="Generated size in pixels")
    url: str = Field(description="Original URL")
    expires_at: Optional[str] = Field(None, description="Cache expiration")


class QRCodeFormats(BaseModel):
    """Available QR code formats."""
    formats: List[QRFormat] = Field(description="Supported formats")
    default_format: QRFormat = Field(description="Default format")


class QRCodeSizes(BaseModel):
    """Available QR code sizes."""
    sizes: List[int] = Field(description="Supported sizes")
    default_size: int = Field(description="Default size")
    min_size: int = Field(description="Minimum size")
    max_size: int = Field(description="Maximum size")


class QRCodeGenerationRequest(BaseModel):
    """QR code generation request."""
    url: HttpUrl = Field(description="URL to encode")
    options: Optional[QRCodeOptions] = Field(None, description="Generation options")


class QRCodeValidationResult(BaseModel):
    """QR code options validation result."""
    valid: bool = Field(description="Whether options are valid")
    errors: Optional[List[str]] = Field(None, description="Validation errors")
    warnings: Optional[List[str]] = Field(None, description="Validation warnings")


class QRCodeBrandingOptions(BaseModel):
    """QR code branding options."""
    logo: Optional[str] = Field(None, description="Base64 encoded logo")
    logo_size: Optional[int] = Field(None, ge=5, le=30, description="Logo size as percentage")
    background_color: Optional[str] = Field(None, description="Background color")
    foreground_color: Optional[str] = Field(None, description="Foreground color")
    border_radius: Optional[int] = Field(None, ge=0, le=50, description="Border radius")
    gradient: Optional[Dict[str, str]] = Field(None, description="Gradient configuration")
    
    @validator('background_color', 'foreground_color')
    def validate_colors(cls, v):
        """Validate color hex codes."""
        if v is not None:
            import re
            if not re.match(r'^#[0-9A-Fa-f]{6}$', v):
                raise ValueError('Color must be a valid hex code')
        return v


class QRCodeBatchRequest(BaseModel):
    """Batch QR code generation request."""
    requests: List[QRCodeGenerationRequest] = Field(description="QR code requests")
    common_options: Optional[QRCodeOptions] = Field(None, description="Common options for all")


class QRCodeBatchResponse(BaseModel):
    """Batch QR code generation response."""
    qr_codes: List[QRCodeResponse] = Field(description="Generated QR codes")
    total: int = Field(description="Total number generated")
    failed: int = Field(description="Number of failures")
    errors: Optional[List[str]] = Field(None, description="Error messages")


class QRCodeAnalytics(BaseModel):
    """QR code scan analytics."""
    total_scans: int = Field(description="Total number of scans")
    scans_today: int = Field(description="Scans today")
    scans_by_date: List[Dict[str, Any]] = Field(description="Scans by date")
    top_devices: List[Dict[str, Any]] = Field(description="Top devices")
    top_countries: List[Dict[str, Any]] = Field(description="Top countries")
    scan_locations: List[Dict[str, Any]] = Field(description="Scan locations")


class QRCodeMetadata(BaseModel):
    """QR code metadata."""
    short_code: str = Field(description="Short URL code")
    original_url: HttpUrl = Field(description="Original URL")
    title: Optional[str] = Field(None, description="URL title")
    created_at: str = Field(description="Creation timestamp")
    expires_at: Optional[str] = Field(None, description="Expiration timestamp")
    is_active: bool = Field(description="Whether URL is active")
    qr_code_generated: bool = Field(description="Whether QR code was generated")
    total_scans: int = Field(description="Total number of scans")


class QRCodeTrackingParams(BaseModel):
    """QR code tracking parameters."""
    utm_source: Optional[str] = Field(None, description="UTM source")
    utm_medium: Optional[str] = Field(None, description="UTM medium")
    utm_campaign: Optional[str] = Field(None, description="UTM campaign")
    utm_term: Optional[str] = Field(None, description="UTM term")
    utm_content: Optional[str] = Field(None, description="UTM content")
    custom_params: Optional[Dict[str, str]] = Field(None, description="Custom parameters")


class QRCodeTemplate(BaseModel):
    """QR code template."""
    name: str = Field(description="Template name")
    description: str = Field(description="Template description")
    options: QRCodeOptions = Field(description="Default options")
    branding: Optional[QRCodeBrandingOptions] = Field(None, description="Branding options")
    use_case: str = Field(description="Intended use case")


class QRCodeDataValidation(BaseModel):
    """QR code data validation result."""
    is_valid: bool = Field(description="Whether data is valid")
    data_type: str = Field(description="Type of data (url, text, email, etc.)")
    data: str = Field(description="Validated data")
    warnings: Optional[List[str]] = Field(None, description="Validation warnings")


class QRCodeUsageStats(BaseModel):
    """QR code usage statistics."""
    total_generated: int = Field(description="Total QR codes generated")
    generated_today: int = Field(description="Generated today")
    generated_this_week: int = Field(description="Generated this week")
    generated_this_month: int = Field(description="Generated this month")
    most_popular_format: QRFormat = Field(description="Most popular format")
    most_popular_size: int = Field(description="Most popular size")
    average_scans_per_code: float = Field(description="Average scans per QR code")

File: sdk/javascript/src/services/qr.ts
import { HTTPClient } from '../http-client';
import { API_ENDPOINTS } from '../config';
import { QRCodeOptions, QRCodeResponse, RequestConfig } from '../types';

export interface QRCodeFormats {
  formats: string[];
  defaultFormat: string;
}

export interface QRCodeSizes {
  sizes: number[];
  defaultSize: number;
  minSize: number;
  maxSize: number;
}

export interface QRCodeGenerationRequest {
  url: string;
  options?: QRCodeOptions;
}

export interface QRCodeValidationResult {
  valid: boolean;
  errors?: string[];
  warnings?: string[];
}

export class QRService {
  constructor(private client: HTTPClient) {}

  /**
   * Get available QR code formats
   */
  async getFormats(config?: RequestConfig): Promise<QRCodeFormats> {
    return this.client.get<QRCodeFormats>(API_ENDPOINTS.QR_FORMATS, config);
  }

  /**
   * Get available QR code sizes
   */
  async getSizes(config?: RequestConfig): Promise<QRCodeSizes> {
    return this.client.get<QRCodeSizes>(API_ENDPOINTS.QR_SIZES, config);
  }

  /**
   * Validate QR code options
   */
  async validateOptions(
    options: QRCodeOptions,
    config?: RequestConfig
  ): Promise<QRCodeValidationResult> {
    return this.client.post<QRCodeValidationResult>(
      API_ENDPOINTS.QR_VALIDATE,
      options,
      config
    );
  }

  /**
   * Get QR code preview without saving
   */
  async getPreview(
    url: string,
    options?: QRCodeOptions,
    config?: RequestConfig
  ): Promise<QRCodeResponse> {
    return this.client.post<QRCodeResponse>(
      API_ENDPOINTS.QR_PREVIEW,
      { url, options },
      config
    );
  }

  /**
   * Generate QR code for any URL
   */
  async generate(
    request: QRCodeGenerationRequest,
    config?: RequestConfig
  ): Promise<QRCodeResponse> {
    return this.client.post<QRCodeResponse>(
      API_ENDPOINTS.QR_GENERATE,
      request,
      config
    );
  }

  /**
   * Generate QR code for a short URL by short code
   */
  async generateForURL(
    shortCode: string,
    options?: QRCodeOptions,
    config?: RequestConfig
  ): Promise<QRCodeResponse> {
    const params = new URLSearchParams();
    if (options?.size) params.set('size', options.size.toString());
    if (options?.format) params.set('format', options.format);
    if (options?.level) params.set('level', options.level);
    if (options?.margin) params.set('margin', options.margin.toString());
    if (options?.darkColor) params.set('darkColor', options.darkColor);
    if (options?.lightColor) params.set('lightColor', options.lightColor);

    const url = `${API_ENDPOINTS.QR_FOR_URL(shortCode)}${
      params.toString() ? `?${params.toString()}` : ''
    }`;
    
    return this.client.get<QRCodeResponse>(url, config);
  }

  /**
   * Download QR code as file
   */
  async download(
    shortCode: string,
    options?: QRCodeOptions & { filename?: string },
    config?: RequestConfig
  ): Promise<Blob> {
    const qrResponse = await this.generateForURL(shortCode, options, config);
    
    // Convert base64 to blob for download
    if (qrResponse.format === 'svg') {
      return new Blob([qrResponse.data], { type: 'image/svg+xml' });
    } else {
      // Handle base64 PNG data
      const binaryString = atob(qrResponse.data.replace(/^data:image\/png;base64,/, ''));
      const bytes = new Uint8Array(binaryString.length);
      for (let i = 0; i < binaryString.length; i++) {
        bytes[i] = binaryString.charCodeAt(i);
      }
      return new Blob([bytes], { type: 'image/png' });
    }
  }

  /**
   * Generate QR code with custom branding
   */
  async generateBranded(
    url: string,
    options: QRCodeOptions & {
      logo?: string; // Base64 encoded logo
      logoSize?: number; // Logo size as percentage of QR code
      backgroundColor?: string;
      foregroundColor?: string;
      borderRadius?: number;
    },
    config?: RequestConfig
  ): Promise<QRCodeResponse> {
    return this.client.post<QRCodeResponse>(
      `${API_ENDPOINTS.QR_GENERATE}/branded`,
      { url, options },
      config
    );
  }

  /**
   * Generate batch QR codes for multiple URLs
   */
  async generateBatch(
    requests: QRCodeGenerationRequest[],
    config?: RequestConfig
  ): Promise<QRCodeResponse[]> {
    const response = await this.client.post<{ qrCodes: QRCodeResponse[] }>(
      `${API_ENDPOINTS.QR_GENERATE}/batch`,
      { requests },
      config
    );
    return response.qrCodes;
  }

  /**
   * Get QR code analytics (if URL belongs to user)
   */
  async getAnalytics(
    shortCode: string,
    period?: '24h' | '7d' | '30d' | '90d',
    config?: RequestConfig
  ): Promise<{
    totalScans: number;
    scansToday: number;
    scansByDate: { date: string; scans: number }[];
    topDevices: { device: string; scans: number }[];
    topCountries: { country: string; scans: number }[];
  }> {
    const params = new URLSearchParams();
    if (period) params.set('period', period);

    const url = `${API_ENDPOINTS.QR_FOR_URL(shortCode)}/analytics${
      params.toString() ? `?${params.toString()}` : ''
    }`;
    
    return this.client.get(url, config);
  }

  /**
   * Create QR code with tracking parameters
   */
  async generateWithTracking(
    url: string,
    trackingParams: { [key: string]: string },
    options?: QRCodeOptions,
    config?: RequestConfig
  ): Promise<QRCodeResponse> {
    // Add tracking parameters to URL
    const urlObj = new URL(url);
    Object.entries(trackingParams).forEach(([key, value]) => {
      urlObj.searchParams.set(key, value);
    });

    return this.generate(
      {
        url: urlObj.toString(),
        options,
      },
      config
    );
  }

  /**
   * Get QR code metadata
   */
  async getMetadata(
    shortCode: string,
    config?: RequestConfig
  ): Promise<{
    shortCode: string;
    originalURL: string;
    title?: string;
    createdAt: string;
    expiresAt?: string;
    isActive: boolean;
    qrCodeGenerated: boolean;
    totalScans: number;
  }> {
    return this.client.get(`${API_ENDPOINTS.QR_FOR_URL(shortCode)}/metadata`, config);
  }

  /**
   * Validate QR code data
   */
  validateQRData(data: string): {
    isValid: boolean;
    type: 'url' | 'text' | 'email' | 'phone' | 'sms' | 'wifi' | 'other';
    data: string;
  } {
    // Basic URL validation
    try {
      new URL(data);
      return { isValid: true, type: 'url', data };
    } catch {
      // Check for other formats
      if (data.startsWith('mailto:')) {
        return { isValid: true, type: 'email', data };
      }
      if (data.startsWith('tel:')) {
        return { isValid: true, type: 'phone', data };
      }
      if (data.startsWith('sms:')) {
        return { isValid: true, type: 'sms', data };
      }
      if (data.startsWith('WIFI:')) {
        return { isValid: true, type: 'wifi', data };
      }
      
      return { isValid: true, type: 'text', data };
    }
  }

  /**
   * Get default QR code options
   */
  getDefaultOptions(): QRCodeOptions {
    return {
      size: 200,
      format: 'png',
      level: 'M',
      margin: 4,
      darkColor: '#000000',
      lightColor: '#FFFFFF',
    };
  }

  /**
   * Merge options with defaults
   */
  mergeOptions(options?: QRCodeOptions): QRCodeOptions {
    return {
      ...this.getDefaultOptions(),
      ...options,
    };
  }
}